### PR TITLE
Dev1

### PR DESCRIPTION
**Pull request by fluckmic**

### CODE REVIEW

**Comment by fluckmic** [comment]
I hope you get a notification if i mention you here, @stephanemagnenat.

**Comment by stephanemagnenat** [comment]
A general comment: as the this library's QML code is originally indented by tabs, your changes should follow the same style.

File: AdaHouse.qml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        Transform {
+            matrix: t.inverted()
+        }
+    ]
+
+    Entity {
+
+        components: [
+            SceneLoader {
+                source: "/models/bluehouse.qgltf"       // at the moment i use the 3d model of the blue house for the ada house
```
Comment (stephanemagnenat) [inline AdaHouse.qml]: Why? I'm sure there is a good reason so it will be helpful to write it down.

File: BlueHouse.qml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    ]
+
+    Entity {
+
+        components: [
+            SceneLoader {
+                source: "/models/bluehouse.qgltf"
+            },
+            Transform {
+                //rotation: fromAxisAndAngle(Qt.vector3d(0,0,1), 180) // -> necessary?
```
Comment (stephanemagnenat) [inline BlueHouse.qml]: If the model works correctly without the transform, please remove it.

File: OrangeHouse.qml [hunk 3]
```diff
@@ -1,14 +1,27 @@
...
 Entity {
-	components: [
-		SceneLoader {
-			source: "/models/orangehouse.qgltf"
-		},
-		Transform {
-			translation: Qt.vector3d(0, 0, -0.038) // the marker is on the top of the house
-		}
-	]
+    property matrix4x4 t
```
Comment (stephanemagnenat) [inline OrangeHouse.qml]: It would be more helpful to name the transform with its semantic meaning, for instance worldPose.

File: OrangeHouse.qml [hunk 4]
```diff
@@ -1,14 +1,27 @@
...
+    ]
+
+    Entity {
+
+        components: [
+            SceneLoader {
+                source: "/models/orangehouse.qgltf"
+            },
+            Transform {
+                rotation: fromAxisAndAngle(Qt.vector3d(0,0,1), 180) // buggy 3d model?
```
Comment (stephanemagnenat) [inline OrangeHouse.qml]: Initially that was not necessary, so I am not sure the 3D model is buggy, maybe the world to camera transform has a problem?

File: android/AndroidManifest.xml [hunk 5]
```diff
@@ -42,7 +42,7 @@
             <!-- Background running -->
         </activity>
     </application>
-    <uses-sdk android:minSdkVersion="16" android:targetSdkVersion="23"/>
+    <uses-sdk android:minSdkVersion="17" android:targetSdkVersion="23"/>
```
Comment (stephanemagnenat) [inline android/AndroidManifest.xml:45]: Why do we now need version 17 of the SDK?

File: main.qml [hunk 6]
```diff
@@ -52,68 +151,146 @@ ApplicationWindow {
...
 
-		captureMode: Camera.CaptureViewfinder
+        captureMode: Camera.CaptureViewfinder
 		cameraState: Camera.LoadedState
-		//deviceId: QtMultimedia.availableCameras[1].deviceId // hack to use second camera on laptop
+
+        //deviceId: QtMultimedia.availableCameras[1].deviceId // hack to use second camera on laptop
 	}
 
+    // HACK
```
Comment (stephanemagnenat) [inline main.qml]: Why? If it is a hack, why is it still here? If it works around a bug in Qt, then please explain in detail why this is here.

File: main.qml [hunk 7]
```diff
@@ -52,68 +151,146 @@ ApplicationWindow {
...
+    Timer {
+        running: true
+        interval: 3000
+        onTriggered: {
+            camera.stop();
+            camera.start();
+        }
+    }
+
+    // Timer to update model.
```
Comment (stephanemagnenat) [inline main.qml]: Update which model with which information?

File: main.qml [hunk 8]
```diff
@@ -52,68 +151,146 @@ ApplicationWindow {
...
+
+    // For more information check out experimentfilter.cpp
+    ExperimentFilter {
+        id: experimentFilter
+        active: experimentFilterToggle.active
+    }
+
+    Item {
+
+        Connections {
```
Comment (stephanemagnenat) [inline main.qml]: Please try to use a `Repeater` to avoid duplication.

File: markermodel.cpp [hunk 9]
```diff
@@ -0,0 +1,704 @@
...
+
+        return lhs;
+    };
+
+    // ***************************************/
+    // Start of model specific implementation.
+
+    /* For the current implementation of the model the following assumptions are made:
+     *
+     * > The position of the marker relative to eachother is fix.
```
Comment (stephanemagnenat) [inline markermodel.cpp]: "eachother" => "each other"

File: markermodel.h [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+const double thDistanceToLastUpdate = 100;       // in ms
+
+
+class MarkerModelMonitor;
+
+/*****************************
+ * MARKER MODEL              *
+ *****************************/
+
+class MarkerModel : public QObject, public TransMem {
```
Comment (stephanemagnenat) [inline markermodel.h]: What is a marker model? There should be a minimum of documentation here.

File: markermodel.h [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+    // Getter for the QML interface
+    QMatrix4x4 world2cam();
+    QMatrix4x4 world2orangeHouse();
+    QMatrix4x4 world2adaHouse();
+
+    bool world2camActive();
+    bool world2orangeHouseActive();
+    bool world2adaHouseActive();
+
+    // Private members (P for private)
```
Comment (stephanemagnenat) [inline markermodel.h]: Do not encode visibility of members into their names.

File: markermodel.h [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+    QThread monitorThread;
+
+signals:
+
+    // Signals for the monitor.
+
+    // Object emits this signal everytime it received an update for a tranformation through
+    // updateLinkNow(..) from the QML model to inform the monitor also about the update.
+    void linkUpdate(const std::string &srcFrame, const std::string &dstFrame, const Timestamp &ts,
+                    const qPair &transf, const float &conf);
```
Comment (stephanemagnenat) [inline markermodel.h]: Do not pass const reference to basic types, just pass by copy.

File: markermodel.h [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+
+/*****************************
+ * TRANSFORMATION UPDATE     *
+ *****************************/
+
+// Basic type of an analysis.
+struct Analysis{
+
+    // Don't allow creation of an object of this type.
+    virtual ~Analysis(){}
```
Comment (stephanemagnenat) [inline markermodel.h]: Do not write an empty destructor like this, just tell the compiler to use the default one: `virtual ~Analysis() = default;`.

File: markermodel.h [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+    Timestamp tZero;
+    // Container for all the results.
+    std::list<AnalysisSingleResult> results;
+};
+
+/************************************************
+ * LINK UPDATE ANALYSIS *
+ ************************************************/
+
+/* This analysis compares the for each link update, the transformation with the transformation
```
Comment (stephanemagnenat) [inline markermodel.h]: "This analysis compares the for [...]", the what?

**Review by stephanemagnenat — CHANGES_REQUESTED** [review]
Besides the specific issues I pointed out in the code, there is a general design problem with MarkeModel, as there should not be marker-specific code in the C++ part. Rather, there should be C++ classes and functions that are used in the QML code to encode the specific logic of marker fusion. Beside that, the indentation does not match the project one, so this should be corrected.

**Pull request closed without merging**